// @azure/communication-react/dist/dist-esm/react-composites/src/composites/common/ControlBar/CustomButton.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
// eslint-disable-next-line no-restricted-imports
import { Icon, mergeStyleSets } from '@fluentui/react';
import { ControlBarButton } from "../../../../../react-components/src";
import React from 'react';
/**
 * Max number of Custom Buttons in primary and secondary ControlBar
 * Does not include existing buttons in the controlBar.
 *
 * @private
 */
export const CUSTOM_BUTTON_OPTIONS = {
    MAX_PRIMARY_DESKTOP_CUSTOM_BUTTONS: 3,
    MAX_PRIMARY_MOBILE_CUSTOM_BUTTONS: 1,
    MAX_SECONDARY_DESKTOP_CUSTOM_BUTTONS: 2
};
/** @private */
export const generateCustomCallControlBarButton = (onFetchCustomButtonProps, displayType) => {
    const allButtonProps = onFetchCustomButtonProps === null || onFetchCustomButtonProps === void 0 ? void 0 : onFetchCustomButtonProps.map(callback => callback({
        displayType
    }));
    return {
        primary: generateCustomControlBarButtons('primary', allButtonProps),
        secondary: generateCustomControlBarButtons('secondary', allButtonProps),
        overflow: generateCustomControlBarButtons('overflow', allButtonProps)
    };
};
/** @private */
const generateCustomControlBarButtons = (placement, customButtons) => customButtons ? customButtons.filter(buttonProps => buttonProps.placement === placement).map((buttonProps, i) => internalProps => {
    var _a, _b, _c, _d, _e, _f, _g, _h;
    if (buttonProps['onRenderButton']) {
        return buttonProps['onRenderButton'](internalProps);
    }
    return React.createElement(ControlBarButton, { ariaDescription: (_a = buttonProps.ariaDescription) !== null && _a !== void 0 ? _a : internalProps.ariaDescription, ariaLabel: (_b = buttonProps.ariaLabel) !== null && _b !== void 0 ? _b : internalProps.ariaLabel, disabled: (_c = buttonProps.disabled) !== null && _c !== void 0 ? _c : internalProps.disabled, id: (_d = buttonProps.id) !== null && _d !== void 0 ? _d : internalProps.id, key: (_e = buttonProps.key) !== null && _e !== void 0 ? _e : `${buttonProps.placement}_${i}`, onClick: (_f = buttonProps.onItemClick) !== null && _f !== void 0 ? _f : internalProps.onClick, onRenderIcon: () => { var _a, _b, _c; return React.createElement(Icon, { iconName: (_c = (_a = buttonProps.iconName) !== null && _a !== void 0 ? _a : (_b = internalProps.iconProps) === null || _b === void 0 ? void 0 : _b.iconName) !== null && _c !== void 0 ? _c : 'ControlButtonOptions' }); }, showLabel: (_g = buttonProps.showLabel) !== null && _g !== void 0 ? _g : internalProps.showLabel, text: (_h = buttonProps.text) !== null && _h !== void 0 ? _h : internalProps.text, styles: mergeStyleSets(internalProps.styles, buttonProps.styles) });
}) : [];
/** @private */
const generateCustomDrawerButtons = (placement, customButtons) => customButtons ? customButtons.filter(buttonProps => buttonProps.placement === placement).map((buttonProps, i) => (Object.assign(Object.assign({}, buttonProps), { disabled: buttonProps.disabled, iconProps: {
        iconName: buttonProps.iconName
    }, id: buttonProps.id, itemKey: buttonProps.key ? '' + buttonProps.key : `${buttonProps.placement}_${i}`, onItemClick: buttonProps.onItemClick, text: buttonProps.text }))) : [];
/** @private */
export const generateCustomCallDrawerButtons = (onFetchCustomButtonProps, displayType) => {
    const customButtons = onFetchCustomButtonProps === null || onFetchCustomButtonProps === void 0 ? void 0 : onFetchCustomButtonProps.map(callback => callback({
        displayType
    }));
    return {
        primary: generateCustomDrawerButtons('primary', customButtons),
        secondary: generateCustomDrawerButtons('secondary', customButtons),
        overflow: generateCustomDrawerButtons('overflow', customButtons)
    };
};
/** @private */
export const generateCustomCallDesktopOverflowButtons = (onFetchCustomButtonProps, displayType) => {
    const customButtons = onFetchCustomButtonProps === null || onFetchCustomButtonProps === void 0 ? void 0 : onFetchCustomButtonProps.map(callback => callback({
        displayType
    }));
    return {
        primary: generateCustomDrawerButtons('primary', customButtons).map(drawerMenuItemToContextualMenuItem),
        secondary: generateCustomDrawerButtons('secondary', customButtons).map(drawerMenuItemToContextualMenuItem),
        overflow: generateCustomDrawerButtons('overflow', customButtons).map(drawerMenuItemToContextualMenuItem)
    };
};
/** @private */
export const drawerMenuItemToContextualMenuItem = (item) => (Object.assign(Object.assign({}, item), { key: item.itemKey, onClick: item.onItemClick ? ev => {
        var _a;
        (_a = item.onItemClick) === null || _a === void 0 ? void 0 : _a.call(item, ev);
    } : undefined, subMenuProps: item.subMenuProps ? {
        items: item.subMenuProps.map(drawerMenuItemToContextualMenuItem)
    } : undefined }));
/** @private */
export const onFetchCustomButtonPropsTrampoline = (options) => {
    var _a;
    let response = undefined;
    response = (_a = options) === null || _a === void 0 ? void 0 : _a.onFetchCustomButtonProps;
    return response;
};
